db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/FromBaseTable.java

```diff
@@ -1730,6 +1791,28 @@ protected ResultSetNode genProjectRestrict(int numTables)
                 getContextManager());
     }
 
+    private void markFirstColumnReferencedForIndexIteratorMode() throws StandardException {
+        if (numUnusedLeadingIndexFields > 0) {
+            int firstIndexCol = getTrulyTheBestAccessPath().getConglomerateDescriptor().
+                                getIndexDescriptor().baseColumnPositions()[0];
```

Comment:
This probably doesn't work for indexes on expressions since `baseColumnPositions()[0]` may not be the actual first index column. It looks like we want to mark the base table column that is defined as the first column in the index? This might be difficult for an index on expression which, for example, has the first column as `a + b + c`.

I'm not sure in which step this function is called. But in modifying the `FromBaseTable`'s access path to an index on expression, `resultColumns` are also rebuilt to reflect the index columns. If this function is used in cost estimation step, `resultColumns` are not changed yet and it's difficult to associate the two sets of columns.

Reply:
This code is just trying to mark a column as referenced so that it is built into the row during the IndexScan.  It looks like we don't need to do anything here for index on expressions because all of the columns in the index are marked as referenced here:

>         if (irg.isOnExpression()) {
>             assert !oldColumns.isEmpty();
>             DataTypeDescriptor[] indexColumnTypes = irg.getIndexColumnTypes();
>             ValueNode[] exprAsts = irg.getParsedIndexExpressions(getLanguageConnectionContext(), this);
> 
>             for (int i = 0; i < indexColumnTypes.length; i++) {
>                 ResultColumn rc = (ResultColumn) getNodeFactory().getNode(
>                         C_NodeTypes.RESULT_COLUMN,
>                         indexColumnTypes[i],
>                         null,
>                         getContextManager());
>                 rc.setIndexExpression(exprAsts[i]);
>                 rc.setReferenced();

I will update the code to not even enter this path for index on expressions:

>         IndexRowGenerator indexDescriptor = 
>             getTrulyTheBestAccessPath().getConglomerateDescriptor().getIndexDescriptor();
>         if (numUnusedLeadingIndexFields > 0 && !indexDescriptor.isOnExpression()) {
>             int firstIndexCol = getTrulyTheBestAccessPath().getConglomerateDescriptor().
>                                 getIndexDescriptor().baseColumnPositions()[0];